### DISCUSSION TITLE
Retrieve longest stretch of connected atom idx

### DISCUSSION THREAD

**Discussion by agiani99** [rdkit/rdkit#6723] [Q&A]
Hi all,
I have successfully generated an atom_idx_list from mol `smiles = 'Cc1ncsc1-c1ccc([C@H](C)NC(=O)[C@@H]2C[C@@H](O)CN2C(=O)[C@@H](NC(=O)CCCCCCCN(CCOc2ccc(C(=O)c3c(-c4ccc(O)cc4)sc4cc(O)ccc34)cc2)C2CCC2)C(C)(C)C)cc1'` which could be useful to retrieve the central linear portion. `atom_idx_list = [0, 12, 13, 14, 15, 19, 20, 21, 24, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 63, 64, 65, 66]`.
As you can see the list is suboptimal as it still contains single atoms. However, the problem would it be easily solved by a general function that, taking mol and atom_idx_list, will sub-select the longest stretch of connected atoms (i.e. 27,28,29,30,31,32,33,34,35,36,37,63,64,65,66). Any hint for me?
Best
Andrea

**Comment by jrkarki** [comment]
Hi Andrea, 

We already discussed and solved it offline but since there are couple of interests shown by other users, I am giving the solution below.

from rdkit import Chem
from rdkit.Chem.Draw import IPythonConsole
from rdkit.Chem import Draw
IPythonConsole.drawOptions.addAtomIndices = True
IPythonConsole.molSize = 600,600
from rdkit.Chem import rdmolops

smiles = 'Cc1ncsc1-c1ccc([C@H](C)NC(=O)[C@@H]2C[C@@H](O)CN2C(=O)[C@@H](NC(=O)CCCCCCCN(CCOc2ccc(C(=O)c3c(-c4ccc(O)cc4)sc4cc(O)ccc34)cc2)C2CCC2)C(C)(C)C)cc1'

atom_idx_list = [0, 12, 13, 14, 15, 19, 20, 21, 24, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 63, 64, 65, 66]

#since the rdkit cookbook examples use tuples for indexes, I prefer sticking to the common practice 
atom_idx_list = tuple(atom_idx_list)

#the following in-built method does exactly what is needed here 
#Read more at: https://sourceforge.net/p/rdkit/mailman/message/36735316/

index2smiles = Chem.MolFragmentToSmiles(test, atom_idx_list)

#draw isolated structures for the atom_idx_list
index2smiles = Chem.MolFromSmiles(index2smiles)

#next we use rdmolops imported from Chem to find the longest stretch
mol_frags = rdmolops.GetMolFrags(index2smiles, asMols = True)
largest_mol = max(mol_frags, default=index2smiles, key=lambda m: m.GetNumAtoms())
largest_mol_smiles = Chem.MolToSmiles(largest_mol)

Hopefully, we can close the issue now. :)

**Reply by bp-kelley** [reply]
There is, perhaps, a better way to do this.  According to a discussion I had with @adalke many, many years ago, a way of finding the longest path is to start at any atom, go to the atom farthest from that atom (a) and then go to the atom farthest from that (b).  This should be the longest path.

Code:

```
from rdkit import Chem

def _dfs(bond, atom, visited, path, longest):
    """Depth first search to find the longest path from an atom
    Args:
       bond:  the bond we are traversing (None for the first atom)
       atom: the atom we are traversing to
       visited: set of atom indices we have seen on the this path
       path: The list of [(from_bond, to_atom), ...] alreadh in the path
       longest: holds the longest path found so far
    """
    path.append((bond, atom))
    visited.add(atom.GetIdx())
    if not longest or len(path) > len(longest[0]):
        longest.clear()
        longest.append( path.copy() )

    for bond in atom.GetBonds():
        nbr = bond.GetOtherAtom(atom)
        if nbr.GetIdx() not in visited:
            _dfs(bond, nbr, visited, path, longest)

            path.pop()
            visited.remove(nbr.GetIdx())
            
            
def longest_path(mol):
    """Start at any atom (a), go to the atom farthest from that atom (b)
    And then find the atom farthest from (b).  This path is the longest path
    Args:
      mol: the molecule to search
      
    Returns:
       [atom indices for atoms in the longest path], pbond indices for bonds in the longest path]
    """
    for atom in mol.GetAtoms():
        # go to the farthest atom
        visited, path, longest = set(), [], []
        _dfs(None, atom, visited, path, longest)
        
        # from the farthest atom, go to the farthest atom from it
        farthest_atom = longest[0][-1]
        visited, path, longest = set(), [], []
        _dfs(None, atom, visited, path, longest)
        
        # this is the longest path, return
        atom_path = [a.GetIdx() for _,a in longest[0]]
        bond_path = [b.GetIdx() for b,_ in longest[0] if b is not None]
        # set __sssAtoms for jupyter visualization
        m.__sssAtoms = atom_path
        return atom_path, bond_path
```

Example (Note that Jupyter highlights the bonds between selected atoms, so this isn't really the path, but the atoms are, I'll mark these with atom maps for clarity

```
m = Chem.MolFromSmiles("Cc1ncsc1-c1ccc([C@H]NC(=O)[C@@H]2C[C@@H]CN2C(=O)[C@@H]C(C)(C)C)cc1")
atom_path, bond_path = longest_path(m)
for atom_idx in atom_path:
    m.GetAtomWithIdx(atom_idx).SetAtomMapNum(atom_idx+1)
m
```

![image](https://github.com/rdkit/rdkit/assets/9888416/abbc72dd-a9f0-49fe-9a55-6a4681e277ed)

From here we can use a variety of tools like PathToSubMol

```
Chem.PathToSubmol(m, bond_path)
```

![image](https://github.com/rdkit/rdkit/assets/9888416/807c4038-7383-49c6-a8bd-ce670b50001f)

**Reply by agiani99** [reply]
Hi bp-kelley and rkarki.
Thanks for both suggestions. The one from bp-kelley is more general but does not consider the atom_list already given. as the final result is to extract the longest central stretch in compounds like protac (e.g. `smiles = 'Cc1ncsc1-c1ccc([C@H](C)NC(=O)[C@@H]2C[C@@H](O)CN2C(=O)[C@@H](NC(=O)CCCCCCCN(CCOc2ccc(C(=O)c3c(-c4ccc(O)cc4)sc4cc(O)ccc34)cc2)C(C)C)C(C)(C)C)cc1'`) and, in so doing, generating three fragments.
Modifying the Rkarki solution as below, the linker can be extracted in this way:
`from rdkit import Chem
from rdkit.Chem.Draw import IPythonConsole
from rdkit.Chem import Draw
IPythonConsole.drawOptions.addAtomIndices = True
IPythonConsole.molSize = 600,600
from rdkit.Chem import rdmolops

smiles = 'Cc1ncsc1-c1ccc([C@H](C)NC(=O)[C@@H]2C[C@@H](O)CN2C(=O)[C@@H](NC(=O)CCCCCCCN(CCOc2ccc(C(=O)c3c(-c4ccc(O)cc4)sc4cc(O)ccc34)cc2)C(C)C)C(C)(C)C)cc1'

test = Chem.MolFromSmiles(smiles)

atom_idx_list = [0, 12, 13, 14, 15, 19, 20, 21, 24, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 63, 64, 65, 66]

#since the rdkit cookbook examples use tuples for indexes, I prefer sticking to the common practice
atom_idx_list = tuple(atom_idx_list)

#the following in-built method does exactly what is needed here
#Read more at: https://sourceforge.net/p/rdkit/mailman/message/36735316/

index2smiles = Chem.MolFragmentToSmiles(test, atom_idx_list)

#draw isolated structures for the atom_idx_list
index2smiles = Chem.MolFromSmiles(index2smiles)

#next we use rdmolops imported from Chem to find the longest stretch
mol_frags = rdmolops.GetMolFrags(index2smiles, asMols = True)
largest_mol = max(mol_frags, default=index2smiles, key=lambda m: m.GetNumAtoms())
largest_mol_smiles = Chem.MolToSmiles(largest_mol)`

Thanks again.
Andrea
